backlog size aware throttling

pezra opened the pull request:
This PR changes the `retry_in` calculation for concurrency throttled jobs using the `:schedule` strategy. The approach is to estimate when a throttled job will likely be runnable and schedule it for that time. This estimate is based on the current backlog size for the job group (estimated), the expected job runtime and the concurrency limit. 

This should reduce/eliminate unnecessary delays in job processing caused by the current "ignore throttled jobs for TTL seconds" approach while retaining the benefits of the schedule strategy. 


PS: I changed the appraisals file because install_if does all of the bundling work regardless of whether the condition evaluates to true or false. In fact, it fetches the packages list from gems.contribsys.com before it even checks the condition. If BUNDLE_GEMS__CONTRIBSYS__COM isn't set that fetch failed with an auth error.

pezra committed:
schedule throttled jobs based on backlog size

This avoids unnecessary processing delays and smooths the load on
sidekiq workers.

ixti reviewed (commented):
Somehow, I just noticed that I never pressed "submit review" :((

I'm about to release v2.0.0 with drop of Sidekiq-7 and added Sidekiq-8 support. Please rebase on top of main branch after that, and I'll make sure to get to it this week.

pezra committed:
fix lint error

pezra committed:
Add a max_delay to schedule base requeuing

This avoids a failure mode the occurs when there is a large backlog. In
that scenario jobs can get schedule very far in the future even when
they might be processable sooner.

ixti reviewed (approved):
Thank you!
